src/server/replica.cc

```diff
@@ -137,29 +137,39 @@ bool Replica::Start(ConnectionContext* cntx) {
   // 3. Spawn main coordination fiber.
   sync_fb_ = ::boost::fibers::fiber(&Replica::MainReplicationFb, this);
 
+  // 4. Init basic context, which is used only for cancellation.
+  // Full and stable sync context will be used for error propagation as well.
+  cntx_.emplace([this](const auto& ge) {
+    CloseAllSockets();
+    return true;
+  });
+
+  // Lock the running latch.
+  running_latch_mu_.lock();
+
   (*cntx)->SendOk();
   return true;
 }
 
 void Replica::Stop() {
+  // Mark disabled, prevent from retrying.
   if (sock_) {
     sock_->proactor()->Await([this] {
       state_mask_ = 0;  // Specifically ~R_ENABLED.
-      auto ec = sock_->Shutdown(SHUT_RDWR);
-      LOG_IF(ERROR, ec) << "Could not shutdown socket " << ec;
     });
   }
 
-  // Close sub flows.
-  auto partition = Partition(num_df_flows_);
-  shard_set->pool()->AwaitFiberOnAll([&](unsigned index, auto*) {
-    for (auto id : partition[index]) {
-      shard_flows_[id]->Stop();
-    }
-  });
+  // The context if fully reposible for cleaning up on cancellation.
+  {
+    lock_guard lk(cntx_mu_);
+    CHECK(cntx_);
+    cntx_->Cancel();
+  }
 
-  if (sync_fb_.joinable())
-    sync_fb_.join();
+  // Make sure the replica fully stopped and did all cleanups,
+  // so we can freely release resources (connections)
+  running_latch_mu_.lock();
+  running_latch_mu_.unlock();
```

Comment:
Another big issue with having a forcefull stop would be resource cleanup

We have the same issue in the master dflycmd

Once we're done here and closed the socket, it might be dropped (deleted). The coordinator (or a flow) might be stuck somewhere inside a function and access the socket (even just for reads)... but the socket was already freed. So we need to wait for the replica to fully finish all its cleanups

Reply:
and why not `sync_fb_.join()` ?

Reply:
Yeah, that was silly